# Analysis of mobile apps
*This project is about analyzing app data to help developers of a virtual company understand what kinds of apps are likely to attract more users.*
*My goal in this project is to exercise the concepts I've learned so far*

In [1]:
opened_file = open("AppleStore.csv")
from csv import reader
read_file = reader(opened_file)
apps_data_apple = list(read_file)

opened_file = open("googleplaystore.csv")
from csv import reader
read_file = reader(opened_file)
apps_data_google = list(read_file)


In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))


In [3]:
explore_data(apps_data_apple, 0, 5, rows_and_columns=True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7198
Number of columns: 16


In [4]:
explore_data(apps_data_google, 0, 5, rows_and_columns=True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10842
Number of columns: 13


In [5]:
print(apps_data_apple[0])

print(apps_data_google[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [6]:
print(apps_data_google[0])
print(apps_data_google[10473])
print(apps_data_google[10474])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


In [7]:
del apps_data_google[10473]

In [8]:
for app in apps_data_google:
    name = app[0]
    if name == "Instagram":
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [9]:
duplicate_apps = []
unique_apps = []

for app in apps_data_google:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print(len(duplicate_apps))
print(duplicate_apps[:15])
        

1181
['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


**Google Play has 1181 duplicate name entries. As an example, we 
check the name "Quick PDF Scanner + OCR FREE" for duplicates:**

In [10]:
for app in apps_data_google:
    name = app[0]
    if name == "Quick PDF Scanner + OCR FREE":
        print(app)

['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80804', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']


** REmoval of duplicate rows:  we'll only keep the row with the highest number of reviews and remove the other entries for any given app.**

In [11]:
reviews_max={}
for app in apps_data_google[1:]:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews: 
        reviews_max[name] = n_reviews
    if name not in reviews_max:
        reviews_max[name] = n_reviews
        
        
        
        
        

In [12]:
print(len(reviews_max))

9659


In [13]:
android_clean=[]
already_added=[]
for app in apps_data_google[1:]:
    name = app[0]
    n_reviews = float(app[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)








In [14]:
print(len(android_clean))
print(android_clean[0:4])

9659
[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']]


In [15]:
print(android_clean[7940][0])

لعبة تقدر تربح DZ


In [16]:
print(ord("爱"))

29233


In [17]:
def check_string_english(a_string):
    var = 0
    for char in a_string:
        if ord(char) > 127:
            var += 1
    if var > 3:
        return False
    return True

In [18]:
print(check_string_english("Instachat 😜"))
print(check_string_english('Docs To Go™ Free Office Suite'))
print(check_string_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(check_string_english('Docs To Go™ Free Office Suite'))
print(check_string_english("😜"))
print(check_string_english("ab😜xy"))
print(check_string_english("abxy😜😜😜😜"))

True
True
False
True
True
True
False


In [19]:
print(check_string_english('Docs To Go™ Free Office Suite'))
print(check_string_english('Instachat 😜'))
print(check_string_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


In [20]:

## apps_data_apple
## apps_data_google

apple_english = []
for row in apps_data_apple[1:]:
    x = check_string_english(row[1])
    if x == True:
        apple_english.append(row)
    
android_english = []
for row in android_clean[1:]:
    x = check_string_english(row[0])
    if x == True:
        android_english.append(row)

print(len(apple_english))
print(len(android_english))

6183
9613


In [21]:
print(android_english[0])

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


In [22]:
print(android_clean[0])

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


In [23]:
android_english_free = []
for row in android_english:
    if row[6] == "Free":
        android_english_free.append(row)
              
        
print(len(android_english_free))
print(android_english_free[0:2])

8862
[['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']]


In [24]:
apple_english_free = []
for row in apple_english:
    if row[4] == "0.0":
        apple_english_free.append(row)
  
print(len(apple_english_free))
print(apple_english_free[0:2])        
        

3222
[['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']]


In [25]:
print (apple_english_free[0])
print (android_english_free[0])

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']
['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


In [26]:
apple_english_free.insert(0, apps_data_apple[0])
android_english_free.insert(0, apps_data_google[0])

In [27]:
print(apple_english_free[0:2])
print(len(apple_english_free))

print(android_english_free[0:2])
print(len(android_english_free))

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'], ['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']]
3223
[['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']]
8863


In [28]:
def freq_table(dataset, index):
    table = {}
    for row in dataset[1:]:
        x = row[index]
        if x in table:
            table[x] += 1
        else:
            table[x] = 1
    return table

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])       
        

In [29]:
display_table(apple_english_free, 11)


Games : 1874
Entertainment : 254
Photo & Video : 160
Education : 118
Social Networking : 106
Shopping : 84
Utilities : 81
Sports : 69
Music : 66
Health & Fitness : 65
Productivity : 56
Lifestyle : 51
News : 43
Travel : 40
Finance : 36
Weather : 28
Food & Drink : 26
Reference : 18
Business : 17
Book : 14
Navigation : 6
Medical : 6
Catalogs : 4


**The most common genre is "Games". The runner-up genre is "Entertainment". Most of the apps are more for fun than for practical purposes. Based on this frequency table alone, a general recommendation for an app profile can't be given since the number of users of each genre is not known.**

In [30]:
display_table(android_english_free, 9)

Tools : 749
Entertainment : 538
Education : 474
Business : 407
Productivity : 345
Lifestyle : 345
Finance : 328
Medical : 313
Sports : 307
Personalization : 294
Communication : 287
Action : 275
Health & Fitness : 273
Photography : 261
News & Magazines : 248
Social : 236
Travel & Local : 206
Shopping : 199
Books & Reference : 190
Simulation : 181
Dating : 165
Arcade : 164
Video Players & Editors : 157
Casual : 156
Maps & Navigation : 124
Food & Drink : 110
Puzzle : 100
Racing : 88
Role Playing : 83
Libraries & Demo : 83
Auto & Vehicles : 82
Strategy : 80
House & Home : 73
Weather : 71
Events : 63
Adventure : 60
Comics : 54
Beauty : 53
Art & Design : 52
Parenting : 44
Card : 40
Casino : 38
Trivia : 37
Educational;Education : 35
Board : 34
Educational : 33
Education;Education : 30
Word : 23
Casual;Pretend Play : 21
Music : 18
Racing;Action & Adventure : 15
Puzzle;Brain Games : 15
Entertainment;Music & Video : 15
Casual;Brain Games : 12
Casual;Action & Adventure : 12
Arcade;Action & Advent

In [31]:
display_table(android_english_free, 1) 

FAMILY : 1675
GAME : 862
TOOLS : 750
BUSINESS : 407
LIFESTYLE : 346
PRODUCTIVITY : 345
FINANCE : 328
MEDICAL : 313
SPORTS : 301
PERSONALIZATION : 294
COMMUNICATION : 287
HEALTH_AND_FITNESS : 273
PHOTOGRAPHY : 261
NEWS_AND_MAGAZINES : 248
SOCIAL : 236
TRAVEL_AND_LOCAL : 207
SHOPPING : 199
BOOKS_AND_REFERENCE : 190
DATING : 165
VIDEO_PLAYERS : 159
MAPS_AND_NAVIGATION : 124
FOOD_AND_DRINK : 110
EDUCATION : 103
ENTERTAINMENT : 85
LIBRARIES_AND_DEMO : 83
AUTO_AND_VEHICLES : 82
HOUSE_AND_HOME : 73
WEATHER : 71
EVENTS : 63
PARENTING : 58
ART_AND_DESIGN : 56
COMICS : 55
BEAUTY : 53


**The most common genre is "Tools" followed by the genre "Entertainment". Most of the apps are more for fun than for practical purposes. Based on this frequency table alone, a general recommendation for an app profile can't be given since the number of users of each genre is not known.**

In [32]:
freq_table_AppStore = freq_table(apple_english_free, 11)
freq_table_AppStore

{'Book': 14,
 'Business': 17,
 'Catalogs': 4,
 'Education': 118,
 'Entertainment': 254,
 'Finance': 36,
 'Food & Drink': 26,
 'Games': 1874,
 'Health & Fitness': 65,
 'Lifestyle': 51,
 'Medical': 6,
 'Music': 66,
 'Navigation': 6,
 'News': 43,
 'Photo & Video': 160,
 'Productivity': 56,
 'Reference': 18,
 'Shopping': 84,
 'Social Networking': 106,
 'Sports': 69,
 'Travel': 40,
 'Utilities': 81,
 'Weather': 28}

In [33]:
for genre in freq_table_AppStore:
    total = 0
    len_genre = 0
    
    for row in apple_english_free[1:]:
        genre_app = row[11]
        if genre_app == genre:
            user_ratings = float(row[5])
            total += user_ratings
            len_genre += 1
            
    avg = total / len_genre
    print(genre)
    print(avg)
    
    
    
    

Reference
74942.11111111111
Book
39758.5
Business
7491.117647058823
Health & Fitness
23298.015384615384
News
21248.023255813954
Navigation
86090.33333333333
Sports
23008.898550724636
Finance
31467.944444444445
Weather
52279.892857142855
Entertainment
14029.830708661417
Photo & Video
28441.54375
Catalogs
4004.0
Travel
28243.8
Social Networking
71548.34905660378
Lifestyle
16485.764705882353
Utilities
18684.456790123455
Productivity
21028.410714285714
Food & Drink
33333.92307692308
Education
7003.983050847458
Shopping
26919.690476190477
Medical
612.0
Games
22788.6696905016
Music
57326.530303030304


In [34]:
freq_table_GooglePlay = freq_table(android_english_free, 1)
freq_table_GooglePlay

{'ART_AND_DESIGN': 56,
 'AUTO_AND_VEHICLES': 82,
 'BEAUTY': 53,
 'BOOKS_AND_REFERENCE': 190,
 'BUSINESS': 407,
 'COMICS': 55,
 'COMMUNICATION': 287,
 'DATING': 165,
 'EDUCATION': 103,
 'ENTERTAINMENT': 85,
 'EVENTS': 63,
 'FAMILY': 1675,
 'FINANCE': 328,
 'FOOD_AND_DRINK': 110,
 'GAME': 862,
 'HEALTH_AND_FITNESS': 273,
 'HOUSE_AND_HOME': 73,
 'LIBRARIES_AND_DEMO': 83,
 'LIFESTYLE': 346,
 'MAPS_AND_NAVIGATION': 124,
 'MEDICAL': 313,
 'NEWS_AND_MAGAZINES': 248,
 'PARENTING': 58,
 'PERSONALIZATION': 294,
 'PHOTOGRAPHY': 261,
 'PRODUCTIVITY': 345,
 'SHOPPING': 199,
 'SOCIAL': 236,
 'SPORTS': 301,
 'TOOLS': 750,
 'TRAVEL_AND_LOCAL': 207,
 'VIDEO_PLAYERS': 159,
 'WEATHER': 71}

In [39]:
for category in freq_table_GooglePlay:
    total = 0
    len_category = 0
    
    for row in android_english_free[1:]:
        category_app = row[1]
        if category_app == category:
            number_installs = row[5]
            number_installs = number_installs.replace('+', '')
            number_installs = number_installs.replace(',', '')
            number_installs = float(number_installs)
            total += number_installs
            len_category += 1
            
    avg = total / len_category
    print(category)
    print(avg)

AUTO_AND_VEHICLES
647317.8170731707
HOUSE_AND_HOME
1331540.5616438356
ENTERTAINMENT
11640705.88235294
SHOPPING
7036877.311557789
HEALTH_AND_FITNESS
4188821.9853479853
FOOD_AND_DRINK
1924897.7363636363
MAPS_AND_NAVIGATION
4056941.7741935486
PERSONALIZATION
5201482.6122448975
EVENTS
253542.22222222222
GAME
15588015.603248259
BEAUTY
513151.88679245283
LIFESTYLE
1437816.2687861272
LIBRARIES_AND_DEMO
638503.734939759
COMMUNICATION
38456119.167247385
FAMILY
3697848.1731343283
PARENTING
542603.6206896552
PRODUCTIVITY
16787331.344927534
WEATHER
5074486.197183099
COMICS
817657.2727272727
SOCIAL
23253652.127118643
FINANCE
1387692.475609756
TOOLS
10801391.298666667
VIDEO_PLAYERS
24727872.452830188
PHOTOGRAPHY
17840110.40229885
MEDICAL
120550.61980830671
BUSINESS
1712290.1474201474
BOOKS_AND_REFERENCE
8767811.894736841
TRAVEL_AND_LOCAL
13984077.710144928
EDUCATION
1833495.145631068
ART_AND_DESIGN
2021626.7857142857
SPORTS
3638640.1428571427
NEWS_AND_MAGAZINES
9549178.467741935
DATING
854028.830303

In [38]:
print(android_english_free[0:2])

[['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']]
